[wasm] Dead code removal, minor fixes

pavelsavara opened the pull request:
- removed `IJSObject`, `HostObject`, `CoreObject` empty private classes
- removed `TypedArray, Float32Array, Float64Array, Int8Array, Int16Array, Int32Array, Uint16Array, Uint32Array, Uint8ClampedArray, SharedArrayBuffer, Map` unused private classes and their tests
- treat `jsHandle` and `gcHandle` as `IntPtr` on C# icall definition
- moved Length property to `Array` and `Uint8Array` types
- removed unused eventListener helpers and it's tests
- added `Module.stackSave`, `Module.stackRestore`, `Module.stackAlloc` imports
- fixed promise reject signature
- fixed cwrap of `mono_wasm_invoke_method` and `mono_wasm_try_unbox_primitive_and_get_type`
- fixed string interning on `StringDecoder` path
- improved assert helper
- fix for GCHandle identity cache

pavelsavara commented:
/azp run runtime-wasm

kg commented:
Do we need to rename every occurrence of JSObject instead of move it into a separate namespace? It creates a huge amount of diff noise and is going to introduce merge conflicts.

kg commented:
If we're going to change the handles from int to intptr, it might be better to just define a unique type for each category of handle, intptr-style. It can just be a struct containing a single value. Then the type information actually helps us.

pavelsavara commented:
Moved the `JSObject` to `Private` namespace instead, thanks. The `JSHandle` type would need more work, I will skip that now.

pavelsavara committed:
- treat jsHandle and gcHandle as IntPtr on C# icall definition
- removed IJSObject, HostObject, CoreObject empty private classes
- removed Float32Array, Float64Array, Int8Array, Int16Array, Int32Array, Uint16Array, Uint32Array, Uint8ClampedArray, SharedArrayBuffer, Map unused private classes and their tests
- moved Length property to Array and TypedArray types
- removed unused eventListener helpers and it's tests
- made JSObjectPrivate.IsDisposed() method from property in preparation for extension method on public API type
- added Module.stackSave, Module.stackRestore, Module.stackAlloc imports
- fixed promise reject signature
- fixed cwrap of mono_wasm_invoke_method and mono_wasm_try_unbox_primitive_and_get_type
- fixed string interning on StringDecoder path
- improved assert helper

pavelsavara commented:
/azp run runtime-wasm

kg commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/System.Private.Runtime.InteropServices.JavaScript.sln:
```diff
@@ -15,6 +15,12 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{F323228E-200
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "gen", "gen", "{53D4886A-6035-458C-834C-D7EDC02A0E26}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "System.Net.Http", "..\System.Net.Http\src\System.Net.Http.csproj", "{BB41B80A-455B-40D2-98A0-6663DDB4C1FD}"
```
Why did the sln change so much?

pavelsavara replied:
I added all the projects which depend on private interop

kg commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSObject.References.cs:
```diff
@@ -12,8 +12,12 @@ public partial class JSObject : SafeHandleMinusOneIsInvalid
-        public bool IsDisposed { get; private set; }
+        public IntPtr JSHandle => handle;
+        private bool _isDisposed;
+        public bool IsDisposed()
```
```public bool IsDisposed { get; private set; }```

kg commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.CS.Owned.cs:
```diff
@@ -7,9 +7,9 @@ namespace System.Runtime.InteropServices.JavaScript
     public static partial class Runtime
     {
-        private static readonly Dictionary<int, WeakReference<JSObject>> _csOwnedObjects = new Dictionary<int, WeakReference<JSObject>>();
+        private static readonly Dictionary<IntPtr, WeakReference<JSObject>> _csOwnedObjects = new Dictionary<IntPtr, WeakReference<JSObject>>();
```
I think this will significantly regress performance unless you make sure to provide a comparer. IIRC the default dictionary comparer for most types is bad, moving from int to IntPtr probably will hit that.

kg commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.JS.Owned.cs:
```diff
@@ -10,7 +10,7 @@ public static partial class Runtime
         private static object JSOwnedObjectLock = new object();
         // we use this to maintain identity of GCHandle for a managed object
-        private static Dictionary<object, int> GCHandleFromJSOwnedObject = new Dictionary<object, int>();
+        private static Dictionary<object, IntPtr> GCHandleFromJSOwnedObject = new Dictionary<object, IntPtr>();
```
I just noticed that we're not doing a reference compare here. Please add a todo comment or fix it (I think the default behavior here will depend on .Equals of the object, but we want object.ReferenceEquals in all cases.)

pavelsavara replied:
great catch!

kg commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/TypedArray.cs:
```diff
@@ -66,40 +66,15 @@ protected TypedArray(ArrayBuffer buffer, int byteOffset) : base(typeof(T).Name,
             {
-                case Int8Array _:
-                    return TypedArrayTypeCode.Int8Array;
                 case Uint8Array _:
```
If we're going to keep this type generic, we might as well keep the support for the various primitive element types. If we're only keeping Uint8Array we should just rename it and remove the generic parameter entirely.

kg replied:
To keep the code simplification, we can just delete all the specific derived types and keep the base class (making it no longer abstract), and check T at construction time to make sure it's valid. We already have GetTypedArrayType, so we can use that to perform the check.

pavelsavara replied:
makes sense, thanks! I will remove `TypedArray`

kg commented on src/mono/wasm/runtime/cs-to-js.ts:
```diff
@@ -69,7 +69,7 @@ function _unbox_mono_obj_root_with_known_nonprimitive_type_impl(root: WasmRoot<a
         case MarshalType.ARRAY_FLOAT:
         case MarshalType.ARRAY_DOUBLE:
-            throw new Error("Marshalling of primitive arrays are not supported.  Use the corresponding TypedArray instead.");
+            throw new Error("Marshaling of primitive arrays are not supported.  Use the corresponding TypedArray instead.");
```
If we remove the corresponding typed array support we'll need to update this error :) But as mentioned in another comment I think maybe we keep a subset of it.

kg commented on src/mono/wasm/runtime/method-binding.ts:
```diff
@@ -43,11 +42,7 @@ export function _get_class_name(classPtr: MonoClass): string {
 }
 
 export function find_method(klass: MonoClass, name: string, n: number): MonoMethod {
-    const result = cwraps.mono_wasm_assembly_find_method(klass, name, n);
```
Why?

pavelsavara replied:
because otherwise it would capture the lazy function instead of resolved function.

pavelsavara replied:
`_method_descriptions` was not used anywhere

kg replied:
OK, it used to get used for diagnostics purposes, if it's not needed anymore then that's good, we must have refactored it out

pavelsavara committed:
feedeback from @kg

pavelsavara commented:
/azp run runtime-wasm

danmoseley commented:
@pavelsavara there are merge conflicts

lewing commented:
pavel is out this week

pavelsavara committed:
Merge branch 'main' into wasm_interop_prep1

# Conflicts:
#	src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/TypedArray.cs
#	src/mono/wasm/wasm.proj

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
I tested that Blazor demo works with this.
